src/note/note_type.rs

```diff
@@ -0,0 +1,113 @@
+use group::GroupEncoding;
+use halo2_proofs::arithmetic::CurveExt;
+use pasta_curves::pallas;
+use std::hash::{Hash, Hasher};
+
+use subtle::{Choice, ConstantTimeEq, CtOption};
+
+use crate::constants::fixed_bases::{VALUE_COMMITMENT_PERSONALIZATION, VALUE_COMMITMENT_V_BYTES};
+use crate::keys::IssuerValidatingKey;
+
+/// Note type identifier.
+#[derive(Clone, Copy, Debug, Eq)]
+pub struct NoteType(pallas::Point);
+
+pub const MAX_ASSET_DESCRIPTION_SIZE: usize = 512;
+
+// the hasher used to derive the assetID
+#[allow(non_snake_case)]
+fn assetID_hasher(msg: Vec<u8>) -> pallas::Point {
+    // TODO(zsa) replace personalization
+    pallas::Point::hash_to_curve(VALUE_COMMITMENT_PERSONALIZATION)(&msg)
+}
+
+impl NoteType {
+    /// Deserialize the note_type from a byte array.
+    pub fn from_bytes(bytes: &[u8; 32]) -> CtOption<Self> {
+        pallas::Point::from_bytes(bytes).map(NoteType)
+    }
+
+    /// Serialize the note_type to its canonical byte representation.
+    pub fn to_bytes(self) -> [u8; 32] {
+        self.0.to_bytes()
+    }
+
+    /// $DeriveNoteType$.
+    ///
+    /// Defined in [Zcash Protocol Spec § TBD: Note Types][notetypes].
+    ///
+    /// [notetypes]: https://zips.z.cash/protocol/nu5.pdf#notetypes
+    #[allow(non_snake_case)]
+    pub fn derive(ik: &IssuerValidatingKey, asset_desc: &str) -> Self {
+        assert!(!asset_desc.is_empty() && asset_desc.len() <= MAX_ASSET_DESCRIPTION_SIZE);
+
+        let mut s = vec![];
+        s.extend(ik.to_bytes());
+        s.extend(asset_desc.as_bytes());
+
+        NoteType(assetID_hasher(s))
```

Comment:
Use a snake-case name, e.g. `asset_id_hasher`, and remove the clippy lint suppression.